src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,146 @@
+      Function(name, args, results, pres, posts, body map (computeNewBody(args, _)))(f.info)
+
+    // same as functions
+    case m@Method(receiver, name, args, results, pres, posts, body) =>
+      Method(receiver, name, args, results, pres, posts, body map (computeNewBody(args, _)))(m.info)
+
+    // Adds pre-conditions stating the bounds of each argument and a post-condition to check if the body expression
+    // overflows
+    case f@PureFunction(name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPres(args)
+        val newPost = posts ++ getPureBlockPosts(expr, results)
+        PureFunction(name, args, results, newPre, newPost, body)(f.info)
+      case None => f
+    }
+
+    // Same as pure functions
+    case m@PureMethod(receiver, name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPres(args)
+        val newPost = posts ++ getPureBlockPosts(expr, results)
+        PureMethod(receiver, name, args, results, newPre, newPost, body)(m.info)
+      case None => m
+    }
+
+    /* As discussed on the Gobra meeting (27/10/2020), overflow checks should not be added to predicates, assertions
+    * and any other purely logical (i.e. non-executable code) statements and expressions. This seems to be the approach taken
+    * by other verification tools such as FramaC, as noted by Wytse
+    */
+
+    case x => x
+  }
+
+  /**
+    * Adds overflow checks to the bodies of a block. If block is parametrized,
+    * `computeNewBody` adds assumptions about the parameters' bounds
+    */
+  private def computeNewBody(args: Vector[Parameter.In], body: Block): Block = {
+    val assumeArgBounds = args map { case p@Parameter.In(_, typ) => Assume(assertionExprInBounds(p, typ)(p.info))(p.info) }
+    val blockStmts = body.stmts map stmtTransform
+    Block(body.decls, assumeArgBounds ++ blockStmts)(body.info)
+  }
+
+  /**
+    * Computes the pre-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that each argument is within the bounds for its type
+    */
+  private def getPureBlockPres(args: Vector[Parameter.In]): Vector[Assertion] = {
+    args.map { p: Parameter.In => assertionExprInBounds(p, p.typ)(p.info) }
+  }
+
+  /**
+    * Computes the post-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that the expression result is within the bounds of its type
+    */
+  private def getPureBlockPosts(body: Expr, results: Vector[Parameter.Out]): Vector[Assertion] = {
+    // relies on the current assumption that pure functions and methods must have exactly one result argument
+    if (results.length != 1) violation("Pure functions and methods must have exactly one result argument")
+    Vector(assertionExprInBounds(body, results(0).typ)(addAnnotation(body.info)))
+  }
+
+  private def stmtTransform(stmt: Stmt): Stmt = stmt match {
```

Comment:
The other discussion got kind of hijacked, so I am continuing it here.
Only checking some statement does not really look like a concistent approach to me. How do you know that all statements are covered? What is with calls without an assignemnt, e.g. `x, y := 20200202020, 20020200202; foo(x + y)`? What is with make statements and return statements? What is when we add a new statement?

Also, is it ok to not cover pure functions? You could add an assert expression to the internal representation to handle these cases.

Lastly, and this one is more complicated, what is with conditional code, e.g. `x, y := MAXINT, MAXINT; b := false; z := b ? x + y : 4`?

Reply:
1. You are right, I am going to add checks for the other kinds of statements that are missing.

2. I am covering them already by adding the overflow checks to the post-condition. I did not add them to the body though because it consists of an Option[Expr]

3. This triggers an Overflow Error which is in practice unreachable. I agree that this seems overly conservative but it could also be argued that this expression contains a  sub-expression that causes an overflow. In either case, I can improve this in a future PR

Reply:
So, I updated the program transformer to explicitly match every kind of statement and to generate overflow checks for all "non-logical" statements (i.e. for all statements excluding `Inhale`, `Exhale`, `Assert`, `Assume`, `Fold` and `Unfold`, and also `Return` as it does not contain any expression). If we ever add a new statement, the compiler will let us know that we have a non-exhaustive pattern matching.

I also made some simplifications and detected and fixed a bug in cases like this one
```go
type rect struct {
    width, height int
}

func (r rect) getWidth() int {
	return r.width // was throwing an overflow error even though it does not make sense
}
```

The reason for this is that Gobra was not able to prove that the field was within bounds, even though the type system ensures that it is. I thought about adding axioms to the structures asserting that every integer field access is within bounds however this was not so easy to do given that structs are translated to polymorphic tuples at the Viper level. What I ended up doing is to prove that no overflow occurs in an expression, assuming that the struct fields it accesses are within bounds.